src/main/scala/nl.knaw.dans.easy.archivebag/EasyArchiveBag.scala

```diff
@@ -47,30 +47,36 @@ object EasyArchiveBag extends Bagit5FacadeComponent with DebugEnhancedLogging {
     run.unsafeGetOrThrow
   }
 
-  def run(implicit ps: Parameters): Try[URI] = Try {
-    // TODO: refactor this function not to rely on throwing of exceptions
-    (for {
-      optVersionOfId <- bagFacade.getIsVersionOf(ps.bagDir.toPath)
-      _ <- optVersionOfId.map(createRefBagsTxt).getOrElse(Success(()))
-    } yield ()).get // trigger exception if result is Failure
+  def run(implicit ps: Parameters): Try[URI] = {
+    for {
+      maybeVersionOfId <- bagFacade.getIsVersionOf(ps.bagDir.toPath)
+      _ <- maybeVersionOfId.map(createRefBagsTxt).getOrElse(Success(()))
+      zippedBag <- generateUncreatedTempFile()
+      _ <- zipDir(ps.bagDir, zippedBag)
+      response <- putFile(zippedBag)
+      location <- response.map(handleStorageResponse).tried.flatten
+        .doIfSuccess(_ => zippedBag.delete())
+        .doIfFailure { case _ => zippedBag.delete() }
```

Comment:
Isn't there a kind of `doAlways` ?

Reply:
nope, we don't have that implemented.
note that we created `doIfSuccess` and `doIfFailure` ourselves. We hadn't taken this possibility into account. I'd say we leave it as is.

Reply:
OK